src/sage/rings/polynomial/polynomial_ring.py

```diff
@@ -1448,24 +1481,101 @@ def random_element(self, degree=(-1,2), *args, **kwds):
             else:
                 raise ValueError("No polynomial of degree >= 0 has all coefficients zero")
 
-        # Pick a random degree
-        d = randint(degree[0], degree[1])
-
-        # If degree is -1, return the 0 polynomial
-        if d == -1:
+        if degree == (-1, -1):
             return self.zero()
 
-        # If degree is 0, return a random constant term
-        if d == 0:
-            return self(R._random_nonzero_element(*args, **kwds))
+        # If `monic` is set, zero should be ignored
+        if degree[0] == -1:
+            degree = (0, degree[1])
+            has_zero = not monic
```

Comment:
The `has_zero` variable is not serving any purpose anymore and can be removed in favor of `monic`. This will also simplify the non-monic logic.